fenicsadapter/fenicsadapter.py

```diff
@@ -429,59 +429,61 @@ def _are_connected_by_edge(self, v1, v2):
                     return True
         return False
 
-    def _extract_coupling_boundary_edges(self):
+    def _extract_coupling_boundary_edges(self, id_mapping):
         """Extracts edges of mesh which lie on the boundary.
         :return: two arrays of vertex IDs. Array 1 consists of first points of all edges
         and Array 2 consists of second points of all edges
 
         NOTE: Edge calculation is only relevant in 2D cases.
         """
 
-        n = 0
         vertices = dict()
 
         for v1 in dolfin.vertices(self._mesh_fenics):
             if self._coupling_subdomain.inside(v1.point(), True):
                 vertices[v1] = []
-                n += 1
 
         for v1 in vertices.keys():
             for v2 in vertices.keys():
                 if self._are_connected_by_edge(v1, v2):
                     vertices[v1] = v2
                     vertices[v2] = v1
 
-        vertices_1 = []
-        vertices_2 = []
+        vertices1_ids = []
+        vertices2_ids = []
 
         for v1, v2 in vertices.items():
-            vertices_1.append(v1.x(0))
-            vertices_1.append(v1.x(1))
-            vertices_2.append(v2.x(0))
-            vertices_2.append(v2.x(1))
-
-        vertices_1 = np.array(vertices_1)
-        vertices_2 = np.array(vertices_2)
-        vertices1_ids = np.zeros(n)
-        vertices2_ids = np.zeros(n)
+            if v1 is not v2:
+                vertices1_ids.append(id_mapping[(v1.x(0), v1.x(1))])
+                vertices2_ids.append(id_mapping[(v2.x(0), v2.x(1))])
 
-        self._interface.get_mesh_vertex_ids_from_positions(self._mesh_id, n, vertices_1, vertices1_ids)
-        self._interface.get_mesh_vertex_ids_from_positions(self._mesh_id, n, vertices_2, vertices2_ids)
+        vertices1_ids = np.array(vertices1_ids)
+        vertices2_ids = np.array(vertices2_ids)
 
         return vertices1_ids, vertices2_ids
 
-    def set_coupling_mesh(self, mesh, subdomain, use_nearest_projection=False):  # as soon as issue https://github.com/precice/fenics-adapter/issues/53 is fixed change default to use_nearest_projection=True
+    def set_coupling_mesh(self, mesh, subdomain, use_nearest_projection=True):  # as soon as issue https://github.com/precice/fenics-adapter/issues/53 is fixed change default to use_nearest_projection=True
         """Sets the coupling mesh. Called by initalize() function at the
         beginning of the simulation.
         """
         self._coupling_subdomain = subdomain
         self._mesh_fenics = mesh
         self._coupling_mesh_vertices, self._n_vertices = self._extract_coupling_boundary_vertices()
```

Comment:
Also return vertex objects here, might need to add another function `coupling_mesh_vertices = convert(fenics_vertices)` to convert fenics vertices to preCICE vertex coordinate array.